Source/WebCore/Modules/permissions/MainThreadPermissionObserver.h

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "ClientOrigin.h"
+#include "MainThreadPermissionObserverIdentifier.h"
+#include "PermissionDescriptor.h"
+#include "PermissionObserver.h"
+#include "PermissionQuerySource.h"
+#include "PermissionState.h"
+#include "ScriptExecutionContextIdentifier.h"
+
+namespace WebCore {
+
+class Page;
+
+class MainThreadPermissionObserver final : public PermissionObserver {
```

Comment:
Note that I think the design would look a lot nicer if PermissionStatus had a MainThreadPermissionObserver no matter what (even if a main thread object) and if PermissionStatus no longer inherited PermissionObserver.

The current design is a bit awkward but this refactoring doesn't have to be done in this patch.

Reply:
This is a good point--I just did this in a follow up patch. Now there is always a MainThreadPermssionObserver.